### PR TITLE
Added some generic ways to build transactions etc, parameterised by era.

### PR DESCRIPTION
**Pull request by TimSheard**
Added a type (Witness era), and a bunch of operations paremeterised over it.

Introduces era parmetric Fields, and updateer that can apply a field to
an appropriate object from any era. Also introduces the Policy type that
controls how the value in the field is used to update the object.

Using function calls like: newXXX policy eraWitness [field1,..,fieldn]
on can create new objects where the missing fields are given default values.

Details about which fields are Sets, Maps, Sequences, Lists, or Maybe types are abstracted away. Everything is entered as a list, and is implicitly translated to the correct type.

### CODE REVIEW

**Comment by TimSheard** [comment]
You can see some examples of how this is used here:
https://github.com/input-output-hk/cardano-ledger-specs/blob/ts-generic-construction/alonzo/test/lib/Test/Cardano/Ledger/Generic/Props.hs

**Comment by JaredCorduan** [comment]
I have just:
* moved the new modules to `cardano-ledger-test`
* re-written all the examples currently in `Cardano.Ledger.Alonzo.Examples` using the generic updaters.
* used the examples to make a suite of unit tests in `cardano-ledger-test`
* rebased on top of master

I think the examples are now *much* easier to read, and will be easier to maintain.

~I need to remove the examples from `Cardano.Ledger.Alonzo.Examples` still, but I may need help enabling the tests in `cardano-ledger-test` in hydra.~ The old examples are removed, and the new tests are running in Hydra (as evidenced by a CI failure :) )

File: cardano-ledger-test/src/Test/Cardano/Ledger/Generic/Indexed.hs [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+{-
+import Shelley.Spec.Ledger.BaseTypes
+  ( StrictMaybe (..),
+    maybeToStrictMaybe,
+    strictMaybeToMaybe,
+    Network(..),
+    UnitInterval,
+    Nonce (NeutralNonce),
+  )
+-}
```
Comment (JaredCorduan) [inline cardano-ledger-test/src/Test/Cardano/Ledger/Generic/Indexed.hs]: ooops, looks like I forgot some of the clean-up I was going to do. I'll get this soon.

**Review by JaredCorduan — APPROVED** [review]
I think that this is a much better approach to writing examples than we/I were previously using.

**Pull request merged**